pandas/core/reshape/merge.py

```diff
@@ -533,33 +540,37 @@ def merge_asof(
 # TODO: only copy DataFrames when modification necessary
 class _MergeOperation:
     """
-    Perform a database (SQL) merge operation between two DataFrame objects
-    using either columns as keys or their row indexes
+    Perform a database (SQL) merge operation between two DataFrame or Series
+    objects using either columns as keys or their row indexes
     """
 
     _merge_type = "merge"
 
+    indicator_name: Optional[str]
+    left: "DataFrame"
+    right: "DataFrame"
+
     def __init__(
         self,
-        left,
-        right,
-        how="inner",
+        left: FrameOrSeries,
```

Comment:
Haven't debugged to verify but we can mix and match objects here right? So something as follows:

```python
>>> df = pd.DataFrame([[1]])
>>> ser = pd.Series([1], name="a")
>>> pd.merge(df, ser, left_index=True, right_index=True)
   0  a
0  1  1
```

If that's the case then these actually should be `Union[DataFrame, Series]` instead of `FrameOrSeries`. Pretty tricky

Reply:
Correct, left and right can each be either DataFrame or Series, i.e. 4 valid combinations.  Does `FrameOrSeries` assume you're not mix-and-matching?  will update